File: src/lib/sphere.rs
use crate::dot;
use crate::PI;
use crate::{Aabb, HitRecord, Point3, Ray, Vec3};
use crate::{Hittable, Material};

pub struct Sphere<M: Material> {
    center: Point3,
    radius: f64,
    material: M,
}

impl<M: Material> Sphere<M> {
    #[inline]
    pub fn new(cen: Point3, r: f64, material: M) -> Self {
        Self {
            center: cen,
            radius: r,
            material,
        }
    }
}

impl<M: Material> Hittable for Sphere<M> {
    fn hit(&self, r: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord> {
        let oc = r.origin() - self.center;
        let a = r.direction().length_squared();
        let half_b = dot(&oc, &r.direction());
        let c = oc.length_squared() - self.radius.powi(2);
        let discriminant = half_b.powi(2) - a * c;

        if discriminant >= 0. {
            let sqrtd = discriminant.sqrt();
            let mut root = (-half_b - sqrtd) / a;

            if root >= t_min && t_max >= root {
                let p = r.at(root);
                let normal = (p - self.center) / self.radius;
                let theta = -p.y().acos();
                let phi = -p.z().atan2(p.x()) + PI;
                let mut rec = HitRecord::new(
                    p,
                    normal,
                    root,
                    phi / (2. * PI),
                    theta / PI,
                    None,
                    &self.material,
                );

                rec.set_face_normal(r, &normal);

                return Some(rec);
            }

            root = (-half_b + sqrtd) / a;

            if root >= t_min && t_max >= root {
                let p = r.at(root);
                let normal = (p - self.center) / self.radius;
                let theta = -p.y().acos();
                let phi = -p.z().atan2(p.x()) + PI;
                let mut rec = HitRecord::new(
                    p,
                    normal,
                    root,
                    phi / (2. * PI),
                    theta / PI,
                    None,
                    &self.material,
                );

                rec.set_face_normal(r, &normal);

                return Some(rec);
            }
        }

        None
    }

    fn bounding_box(&self, _time0: f64, _time1: f64) -> Option<Aabb> {
        Some(Aabb::new(
            self.center - Vec3::new(self.radius, self.radius, self.radius),
            self.center + Vec3::new(self.radius, self.radius, self.radius),
        ))
    }
}

File: pkg/rtweekend/sphere.go
package rtweekend

import (
	"math"
)

type Sphere struct {
	center   Point3
	radius   float64
	material interface{}
}

func NewSphere(center Point3, radius float64, material Material) *Sphere {
	sphere := new(Sphere)

	sphere.center = center
	sphere.radius = radius
	sphere.material = material

	return sphere
}

func getSphereUV(p *Point3, u, v *float64) {
	theta := math.Acos(-p.Y())
	phi := math.Atan2(-p.Z(), p.X()) + PI

	*u = phi / (2 * PI)
	*v = theta / PI
}

func (s Sphere) Hit(r *Ray, t_min, t_max float64, rec *HitRecord) bool {
	oc := r.Origin().Substract(&s.center)
	a := r.Direction().LengthSquared()
	halfB := Dot(oc, &r.dir)
	c := oc.LengthSquared() - s.radius*s.radius
	discriminant := halfB*halfB - a*c

	setHitRecord := func(distance float64) {
		*rec = HitRecord{}
		rec.t = distance
		rec.p = *r.At(distance)
		normal := rec.p.Substract(&s.center).Divide(s.radius)
		rec.SetFaceNormal(r, normal)
		getSphereUV(normal, &rec.u, &rec.v)
		rec.material = s.material
	}

	if discriminant <= 0 {
		return false
	}

	sqrtd := math.Sqrt(discriminant)
	root := (-halfB - sqrtd) / a

	if (root > t_min) && (t_max > root) {
		setHitRecord(root)

		return true
	}

	root = (-halfB + sqrtd) / a

	if (root > t_min) && (t_max > root) {
		setHitRecord(root)

		return true
	}

	return false
}

func (s Sphere) BoundingBox(time0, time1 float64, outputBox *AABB) bool {
	*outputBox = *NewAABB(
		*s.center.Substract(NewVec3(s.radius, s.radius, s.radius)),
		*s.center.Add(NewVec3(s.radius, s.radius, s.radius)),
	)

	return true
}
